### PR TITLE
Bitexact resize

### PR DESCRIPTION
**Pull request by terfendail**
### This pullrequest changes
Added bit-exact implementation for linear resize

Should be merged with [extra#414](https://github.com/opencv/opencv_extra/pull/414)
Could be merged with [contrib#1482](https://github.com/opencv/opencv_contrib/pull/1482)

### CODE REVIEW

File: modules/dnn/test/test_caffe_importer.cpp [hunk 1]
```diff
@@ -188,7 +191,9 @@ TEST(Reproducibility_SqueezeNet_v1_1, Accuracy)
     Net net = readNetFromCaffe(findDataFile("dnn/squeezenet_v1.1.prototxt", false),
                                findDataFile("dnn/squeezenet_v1.1.caffemodel", false));
 
-    Mat input = blobFromImage(imread(_tf("googlenet_0.png")), 1.0f, Size(227,227), Scalar(), false);
+    Mat sample = imread(_tf("googlenet_0.png"));
+    float resizeFactor = 227 / (float)224; resize(sample, sample, Size(), resizeFactor, resizeFactor, INTER_LINEAR | EVAL_INEXACT);
```
Comment (alalek) [inline modules/dnn/test/test_caffe_importer.cpp]: > (float)224

`244.0f`

File: modules/imgproc/include/opencv2/imgproc.hpp [hunk 2]
```diff
@@ -288,7 +288,11 @@ enum InterpolationFlags{
     - flag is __not__ set: \f$dst( \rho , \phi ) = src(x,y)\f$
     - flag is set: \f$dst(x,y) = src( \rho , \phi )\f$
     */
-    WARP_INVERSE_MAP     = 16
+    WARP_INVERSE_MAP     = 16,
+    /** flag, perform processing without trying to evaluate bit exact result using integral type calculations.
+    Inexact evaluation could be faster on specific platforms but could produce result the slightly differs b/w platforms.
+    */
+    EVAL_INEXACT         = 32
```
Comment (alalek) [inline modules/imgproc/include/opencv2/imgproc.hpp]: I prefer to have `resizeFast()` instead of flag.
Also this would allow to minimize code binary size (like `cvtColor()` size hell).

Or add `resizeBitExact()/cv::bit_exact::resize()` instead because it doesn't support all current resize modes.

File: modules/imgproc/src/resize.cpp [hunk 3]
```diff
@@ -50,12 +50,1326 @@
...
+#include "opencv2/core/hal/intrin.hpp"
 
 #include "opencv2/core/openvx/ovx_defs.hpp"
 #include "resize.hpp"
 
+#include "opencv2/core/softfloat.hpp"
+
 using namespace cv;
 
+#ifndef CV_ALWAYS_INLINE
```
Comment (alalek) [inline modules/imgproc/src/resize.cpp]: Check for existed "CV_INLINE" macro.

File: modules/imgproc/src/resize.cpp [hunk 4]
```diff
@@ -50,12 +50,1326 @@
...
+    CV_ALWAYS_INLINE fixedpoint64(const softdouble& _val) { val = cvRound64(_val * softdouble(fixedOne)); }
+    CV_ALWAYS_INLINE fixedpoint64& operator = (const int8_t& _val) { val = _val * fixedOne; return *this; }
+    CV_ALWAYS_INLINE fixedpoint64& operator = (const int16_t& _val) { val = _val * fixedOne; return *this; }
+    CV_ALWAYS_INLINE fixedpoint64& operator = (const int32_t& _val) { val = _val * fixedOne; return *this; }
+    CV_ALWAYS_INLINE fixedpoint64& operator = (const softdouble& _val) { val = cvRound64(_val * softdouble(fixedOne)); return *this; }
+    CV_ALWAYS_INLINE fixedpoint64& operator = (const fixedpoint64& _val) { val = _val.val; return *this; }
+    template <typename ET>
+    CV_ALWAYS_INLINE fixedpoint64 operator * (const ET& val2) const { return val * val2; } // Wrong rounding is possible for floating point types
+    CV_ALWAYS_INLINE fixedpoint64 operator * (const fixedpoint64& val2) const
+    {
```
Comment (alalek) [inline modules/imgproc/src/resize.cpp]: What is the source of this implementation?

File: modules/imgproc/src/resize.cpp [hunk 5]
```diff
@@ -50,12 +50,1326 @@
...
+class fixedpoint64
+{
+private:
+    static const int fixedShift = 32;
+    static const int64_t fixedOne = (1LL << fixedShift);
+    static const int64_t fixedRound = ((fixedOne >> 1) - 1);
+
+    int64_t val;
+    fixedpoint64(int64_t _val) : val(_val) {}
+    static CV_ALWAYS_INLINE uint64_t fixedround(const uint64_t& _val) { return (_val + fixedRound + ((_val>>fixedShift)&1)); } // banker's rounding
```
Comment (alalek) [inline modules/imgproc/src/resize.cpp]: `fixedround` is already defined as static const.
Static code analyzers dislikes this.

File: modules/imgproc/src/resize.cpp [hunk 6]
```diff
@@ -50,12 +50,1326 @@
...
+
+namespace
+{
+
+class fixedpoint64
+{
+private:
+    static const int fixedShift = 32;
+    static const int64_t fixedOne = (1LL << fixedShift);
+    static const int64_t fixedRound = ((fixedOne >> 1) - 1);
```
Comment (alalek) [inline modules/imgproc/src/resize.cpp]: We still need C++98 code.

File: modules/imgproc/src/resize.cpp (hunk 4 above)
Comment (terfendail) [reply modules/imgproc/src/resize.cpp]: The code was implemented from scratch, but implementation could be influenced by discussions I've read during investigation of existing approaches.

File: modules/imgproc/src/resize.cpp (hunk 3 above)
Comment (terfendail) [reply modules/imgproc/src/resize.cpp]: CV_INLINE doesn't force inlining and could add `static` modificator,

File: apps/createsamples/utility.cpp [hunk 7]
```diff
@@ -49,11 +49,12 @@
 
 #include "utility.hpp"
 #include "opencv2/core.hpp"
-#include "opencv2/core/core_c.h"
-#include "opencv2/imgcodecs/imgcodecs_c.h"
-#include "opencv2/imgproc/imgproc_c.h"
-#include "opencv2/highgui/highgui_c.h"
-#include "opencv2/calib3d/calib3d_c.h"
```
Comment (alalek) [inline apps/createsamples/utility.cpp]: I believe it is better to extract these 600 changes into a separate PR.

File: modules/core/include/opencv2/core/softfloat.hpp [hunk 8]
```diff
@@ -418,12 +421,18 @@ template<> inline short saturate_cast<short>(softdouble a) { return (short)std::
 template<> inline int saturate_cast<int>(softfloat  a) { return cvRound(a); }
 template<> inline int saturate_cast<int>(softdouble a) { return cvRound(a); }
 
-/** @brief Saturate cast to unsigned integer
+template<> inline int64_t saturate_cast<int64_t>(softfloat  a) { return cvRound(a); }
```
Comment (alalek) [inline modules/core/include/opencv2/core/softfloat.hpp]: `cvRound64()` + test for this

File: modules/core/include/opencv2/core/softfloat.hpp [hunk 9]
```diff
@@ -418,12 +421,18 @@ template<> inline short saturate_cast<short>(softdouble a) { return (short)std::
...
+template<> inline int64_t saturate_cast<int64_t>(softdouble a) { return cvRound64(a); }
+
+/** @brief Saturate cast to unsigned integer and unsigned long long integer
 We intentionally do not clip negative numbers, to make -1 become 0xffffffff etc.
 */
 template<> inline unsigned saturate_cast<unsigned>(softfloat  a) { return cvRound(a); }
 template<> inline unsigned saturate_cast<unsigned>(softdouble a) { return cvRound(a); }
 
+template<> inline uint64_t saturate_cast<uint64_t>(softfloat  a) { return cvRound(a); }
```
Comment (alalek) [inline modules/core/include/opencv2/core/softfloat.hpp]: `cvRound64()` + test for this

File: modules/imgproc/test/test_imgwarp.cpp [hunk 10]
```diff
@@ -335,6 +335,30 @@ void CV_ResizeTest::prepare_to_validation( int /*test_case_idx*/ )
...
+CV_ResizeExactTest::CV_ResizeExactTest() : CV_ResizeTest()
+{
+    max_interpolation = 1;
+}
+
+
+void CV_ResizeExactTest::get_test_array_types_and_sizes(int test_case_idx, vector<vector<Size> >& sizes, vector<vector<int> >& types)
+{
+    CV_ResizeTest::get_test_array_types_and_sizes(test_case_idx, sizes, types);
+    interpolation = 5;
```
Comment (alalek) [inline modules/imgproc/test/test_imgwarp.cpp]: Please avoid magic numbers

File: 3rdparty/carotene/src/resize.cpp [hunk 11]
```diff
@@ -106,7 +106,8 @@ bool isResizeLinearOpenCVSupported(const Size2D &ssize, const Size2D &dsize, u32
             && !(ssize.width > 0xffffFFFF || ssize.height > 0xffffFFFF)// Restrict image size since internal index evaluation
                                                                        // is performed with u32
 #endif
-            && dsize.width >= 2 && dsize.height >= 8)
+            && dsize.width >= 2 && dsize.height >= 8
+            && (2*dsize.width != ssize.width || 2*dsize.height != ssize.height)) // 2x downscaling is performed as area in OpenCV wich differs from this implementation
```
Comment (alalek) [inline 3rdparty/carotene/src/resize.cpp]: typo: `which`

File: modules/core/include/opencv2/core/softfloat.hpp [hunk 12]
```diff
@@ -389,6 +389,9 @@ CV_EXPORTS int cvTrunc(const cv::softdouble& a);
 CV_EXPORTS int cvRound(const cv::softfloat&  a);
 CV_EXPORTS int cvRound(const cv::softdouble& a);
 
+/** @brief Rounds a number to nearest even long long integer */
+CV_EXPORTS int64_t cvRound64(const cv::softdouble& a);
```
Comment (alalek) [inline modules/core/include/opencv2/core/softfloat.hpp]: Perhaps we need test for this here: https://github.com/opencv/opencv/blob/master/modules/core/test/test_math.cpp

File: modules/imgproc/src/resize.cpp [hunk 13]
```diff
@@ -50,12 +50,899 @@
...
+private:
+    static const int fixedShift = 32;
+
+    int64_t val;
+    fixedpoint64(int64_t _val) : val(_val) {}
+    static CV_ALWAYS_INLINE uint64_t fixedround(const uint64_t& _val) { return (_val + ((1LL << fixedShift) >> 1)); }
+public:
+    typedef fixedpoint64 WT;
+    CV_ALWAYS_INLINE fixedpoint64() { val = 0; }
+    CV_ALWAYS_INLINE fixedpoint64(const int8_t& _val) { val = _val * (1LL << fixedShift); }
```
Comment (alalek) [inline modules/imgproc/src/resize.cpp]:
```
val = ((int64_t)_val) << fixedShift;
```

File: modules/imgproc/src/resize.cpp [hunk 14]
```diff
@@ -50,12 +50,899 @@
...
+    #else
+        #define CV_ALWAYS_INLINE inline
+    #endif
+#endif
+
+namespace
+{
+
+class fixedpoint64
+{
```
Comment (alalek) [inline modules/imgproc/src/resize.cpp]: I believe these 250 lines with fixedpoint implementation can be moved into separate .inl.hpp file (in `src` directory)

**Comment by vpisarev** [comment]
@terfendail, looks good! Could you squash the commits?

**Comment by terfendail** [comment]
Squashed

**Comment by vpisarev** [comment]
:+1:
